Ghidra/Processors/CP1600/data/languages/CP1600.slaspec

```diff
@@ -0,0 +1,644 @@
+
+jmpdest16:     reloc is address_hi & address_lo [ reloc = ((address_hi << 10) + address_lo) * 2; ] { export *:2 reloc;                         }
+branchdest16:  reloc is branch_sign=0 ; imm10   [ reloc = inst_start + 4 + (imm10 * 2);          ] { export *:2 reloc;                         }
+branchdest16:  reloc is branch_sign=1 ; imm10   [ reloc = inst_start + 4 - (imm10 * 2);          ] { export *:2 reloc;                         }
+addr16:        immvalue is imm10                [ immvalue = imm10 * 2;                          ] { local tmp:2 = immvalue; export tmp;       }
+doubleimm16:   value is value_hi & value_lo     [ value = (value_hi << 8) + value_lo;            ] { local tmp:2 = value & 0xFFFF; export tmp; }
+
+################################################################
+
+macro resultFlags(value) {
+	Z = value == 0;
+	S = value s< 0; 
+}
+
+macro addition(first, second) {
+	local tmpC = carry(first, second);
+	local tmpO = scarry(first, second);
+	first = first + second;
+	C = tmpC;
+	O = tmpO;
+	resultFlags(first);
+}
+
+macro readHalfDoubleWord(reg, dest) {
+	if (reg != R6) goto <read>;
+	reg = reg - 1;
+	<read>
+	dest = (*:2 (reg * 2)) & 0x00FF;
+	if ((reg != R4) || (reg != R5)) goto <done>;
+	reg = reg + 1;
+	<done>
+}
+
+macro readDoubleWord(reg, dest) {
+	local __tmpl__:2 = 0;
+	readHalfDoubleWord(reg, __tmpl__);
+	local __tmph__:2 = 0;
+	readHalfDoubleWord(reg, __tmph__);
+	dest = (__tmph__ << 8) | __tmpl__;
+}
+
+macro comparison(first, second) {
+	local __val__ = first - second;
+	O = sborrow(first, second);
+	C = first < second; 
+	resultFlags(__val__);
+}
+
+macro subtraction(first, second) {
+	local __val__ = first - second;
+	O = sborrow(first, second);
+	C = first < second; 
+	resultFlags(__val__);
+	first = __val__;
+}
+
+################################################################
+
+define pcodeop TerminateCurrentInterrupt;
+define pcodeop SoftwareInterrupt;
+
+################################################################
+
+:ADD addr16, reg0_2 is opcode3_9=0x0058 & reg0_2 ; addr16 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	addition(reg0_2, *:2 addr16);
+}
+
+:ADCR reg0_2 is opcode3_9=0x0005 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	local oldC = zext(C);
+	addition(reg0_2, oldC);
+}
+
+:ADDR reg3_5, reg0_2 is opcode6_9=0x0003 & reg3_5 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	addition(reg0_2, reg3_5);
+}
+
+:AND addr16, reg0_2 is opcode3_9=0x0070 & reg0_2 ; addr16 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	reg0_2 = reg0_2 & (*:2 addr16);
+	resultFlags(reg0_2);
+}
+
+:ANDR reg3_5, reg0_2 is opcode6_9=0x0006 & reg3_5 & reg0_2 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
+	reg0_2 = reg0_2 & reg3_5;
+	resultFlags(reg0_2);
+}
+
+:B branchdest16 is opcode6_9=0x0008 & branch_external=0 & branch_condition=0 ; branchdest16 [ doublebyte=0; globalset(inst_next, doublebyte); ] {
```

Comment:
Oh didn't notice they were in 2 tokens, double check this, but you may need `...` here. maybe something like:

`(opcode6_9=0x0008 & branch_external=0 & branch_condition=0) ... & branchdest16`

Reply:
I think `...` is only needed when the tokens are different sizes. I also think `...` is broken atm as I never got it to work (could have been my implementation though.)